File: backend/app/routers/projects.py
"""Project CRUD API routes.

Provides endpoints for listing, creating, retrieving, and deleting
projects.
"""

from __future__ import annotations

from typing import List

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session

from app.database import get_db
from app.models import Project
from app.schemas import ProjectCreate, ProjectDetailResponse, ProjectResponse

router = APIRouter(prefix="/api/projects", tags=["projects"])


@router.get("", response_model=List[ProjectResponse])
def list_projects(
    skip: int = Query(default=0, ge=0, description="Number of records to skip"),
    limit: int = Query(default=100, ge=1, le=200, description="Max records to return"),
    db: Session = Depends(get_db),
) -> List[ProjectResponse]:
    """Return all projects ordered by creation date descending."""
    projects = (
        db.query(Project)
        .order_by(Project.created_at.desc())
        .offset(skip)
        .limit(limit)
        .all()
    )
    return [ProjectResponse.model_validate(p) for p in projects]


@router.post("", response_model=ProjectResponse, status_code=201)
def create_project(
    payload: ProjectCreate,
    db: Session = Depends(get_db),
) -> ProjectResponse:
    """Create a new project and return it."""
    project = Project(
        name=payload.name,
        description=payload.description or "",
    )
    db.add(project)
    db.commit()
    db.refresh(project)
    return ProjectResponse.model_validate(project)


@router.get("/{project_id}", response_model=ProjectDetailResponse)
def get_project(
    project_id: int,
    db: Session = Depends(get_db),
) -> ProjectDetailResponse:
    """Return a single project by ID including its tasks."""
    project = db.query(Project).filter(Project.id == project_id).first()
    if project is None:
        raise HTTPException(status_code=404, detail="Project not found")
    return ProjectDetailResponse.model_validate(project)


@router.delete("/{project_id}", status_code=204)
def delete_project(
    project_id: int,
    db: Session = Depends(get_db),
) -> None:
    """Delete a project and all its tasks."""
    project = db.query(Project).filter(Project.id == project_id).first()
    if project is None:
        raise HTTPException(status_code=404, detail="Project not found")
    db.delete(project)
    db.commit()
    return None

File: frontend/src/api/projects.js
import apiClient from './client';

/**
 * Fetch all projects.
 *
 * @returns {Promise<Array>} List of project objects.
 */
export async function fetchProjects() {
  const response = await apiClient.get('/api/projects');
  return response.data;
}

/**
 * Fetch a single project by ID.
 *
 * @param {number|string} projectId - The project ID.
 * @returns {Promise<Object>} The project object.
 */
export async function fetchProject(projectId) {
  const response = await apiClient.get(`/api/projects/${projectId}`);
  return response.data;
}

/**
 * Create a new project.
 *
 * @param {{name: string, description?: string}} data - Project creation data.
 * @returns {Promise<Object>} The created project.
 */
export async function createProject(data) {
  const response = await apiClient.post('/api/projects', data);
  return response.data;
}

/**
 * Delete a project by ID.
 *
 * @param {number|string} projectId - The project ID.
 * @returns {Promise<Object>} Confirmation response.
 */
export async function deleteProject(projectId) {
  const response = await apiClient.delete(`/api/projects/${projectId}`);
  return response.data;
}
